### PR TITLE
Some fixes.

### PR DESCRIPTION
**Pull request by matthid**
Fixes a bunch issues I encountered when trying to implement https://github.com/matthid/googleMusicAllAccess.

Note: this includes #194 (because I needed the fix), let me know when you merge #194 so I can rebase and remove the last commit.

### CODE REVIEW

**Comment by vmuriart** [comment]
Do you have tests for the fixes ?

File: setup.py [hunk 1]
```diff
@@ -32,9 +32,7 @@ def _find_msbuild_tool(tool="msbuild.exe", use_windows_sdk=False):
     if use_windows_sdk:
         sdks_root = r"SOFTWARE\Microsoft\Microsoft SDKs\Windows"
         kits_root = r"SOFTWARE\Microsoft\Windows Kits\Installed Roots"
-        kits_suffix = "bin"
-        if PLATFORM == "x64":
-            kits_suffix += r"\x64"
+        kits_suffix = os.path.join("bin",PLATFORM)
```
Comment (denfromufa) [inline setup.py]: this should be a separate pull request, since this is unrelated commit from someone else:

https://github.com/pythonnet/pythonnet/pull/194/files

File: src/runtime/classderived.cs [hunk 2]
```diff
@@ -786,7 +786,7 @@ public static void InvokeCtor(IPythonDerivedType obj, string origCtorName, Objec
                         PyObject[] pyargs = new PyObject[args.Length];
                         for (int i = 0; i < args.Length; ++i)
                         {
-                            pyargs[i] = new PyObject(Converter.ToPython(args[i], args[i].GetType()));
+                            pyargs[i] = new PyObject(Converter.ToPython(args[i], args[i]?.GetType()));
```
Comment (denfromufa) [inline src/runtime/classderived.cs]: I think we are not ready for C# 6+ only support. Please do not use "Null-conditional operators" in C# 4.0 codebase.

**Comment by denfromufa** [comment]
@matthid I left 2 comments in your code, in addition can you please rebase/squash all 4 commits into one logical commit. I agree with @vmuriart that the unit tests would be great to include!

File: setup.py (hunk 1 above)
Comment (denfromufa) [inline setup.py]: note that #194 is linked with other pull requests. @tonyroberts is going to merge these changes once appveyor CI gets fixed.

**Comment by vmuriart** [comment]
From looking at the commit titles, it looks like 3 commits its about right. *Seems* to be addressing 3 different issues. in-lieu of doing 3 separate `pr`, i think 1 with three commits is ok. In case I am wrong, and all three issues are actually the same, then yes should be squashed.

**Comment by denfromufa** [comment]
@vmuriart thanks for correcting me :) @matthid you pull request looks good if #194 gets merged!

only tests are missing - not sure if you can get equivalent C# code for your failing F# code?

**Comment by denfromufa** [comment]
also need to downgrade your syntax to C# 4.0

**Comment by matthid** [comment]
@denfromufa Yes I agree with all points. I just wanted to quickly show up the problems. It could take me a while to clean this up as I'm trying to work on something else (I just figured it would be a good idea to let people know there are some fixes available). If someone wants to step in until then feel free to do so. If you encounter problems while reproducing feel free to ask. I'm pretty sure everything can be reproduced with C# as well.

Is there a strong reason to stay on C# 4?

**Comment by matthid** [comment]
By the way after all these fixes I still see 
```
  clrmodule -> C:\PROJ\googleMusicAllAccess\temp\pythonnet\src\clrmodule\bin\x86\ReleaseWin\clrmodule.dll
C:\PROJ\googleMusicAllAccess\temp\pythonnet\packages\UnmanagedExports.1.2.6\tools\RGiesecke.DllExport.targets(42,5): error : Microsoft.Build.Utilities.ToolLocationHelper could not find ildasm.exe. [C:\PROJ\googleMusicAllAccess\temp\pythonnet\src\clrmodule\clrmodule.csproj]
```

in the build output. So there is still something wrong. As I used a VS developer prompt everything should be available. But I didn't investigate as the build continues and the package is available in `dist/`

**Comment by denfromufa** [comment]
@matthid we are tracking this issue with Unmanaged Exports here:

https://github.com/pythonnet/pythonnet/issues/216

**Comment by denfromufa** [comment]
@matthid c# 6 syntax should be downgraded to c# 4 due to .NET 4.0 and Mono compatibility.
This can be re-considered if someone is willing to look into this closer.

**Comment by matthid** [comment]
Let me first say: I have no problem with downgrading (once I have some time... the project looks a bit dead anyway as there are quite some outstanding PRs). But IMHO those are the wrong reasons. You should ensure mono and net40 compat though build and test.

There is nothing stopping you from compiling c#6 to net40. Only a small subset is not working because of missing .net classes but even those you can get working by defining the missing classes yourself (if really needed). Just google for c#6 and net4 to get sources for this.

The travis build is green, which I would say means it works. Btw later mono versions only run with net45 anyway.

**Comment by denfromufa** [comment]
@matthid I read about partial compatibility of C# 6 with .NET 4.0, but did not know that Mono defaults to C# 6. However looks like their main documentation is outdated, which is where I looked first: http://www.mono-project.com/docs/about-mono/languages/csharp/

C# 6 is not ready in Mono compiler according to that page.

WRT pythonnet slow development cycle - please note that it is still more alive development than IronPython and the codebase is older than both IronPython and F#!

**Comment by matthid** [comment]
@denfromufa Yes Thanks, and the code worked great for me (after the bugfixes at least). I tried IronPython first and couldn't make it do anything :)

Yeah the page is really outdated. Mono moved fast after MS open sourced a lot of things... I think they use Roslyn now and can evolve as fast as Visual Studio...

**Comment by matthid** [comment]
I cherry picked the original commit from #194 this should make it possible for you to merge both PRs, and it makes it easy to use my branch for other things. Just to be sure you should merge #194 first (but git should be able to handle that just fine).

Additionally I added test for 2823941. Now I'm looking into test for 35cd5fa. I think cc8b796 is quite hard to test...

**Comment by matthid** [comment]
I'm not using python that much, but as far as I understand it I had to add a module to be able to use relative imports. I hope its fine the way I did it, please review.

**Comment by matthid** [comment]
Maybe a stupid question: But where are the `Python.Embedding` tests run? It seems like they are not run as part of the test suite?

**Comment by denfromufa** [comment]
@matthid that is a good point! embedding tests are not run on travis or appveyor. Once I did managed to run them locally and passed, but there is something brittle there with importing file paths:

https://github.com/pythonnet/pythonnet/blob/master/src/embed_tests/pyimport.cs#L27

**Comment by denfromufa** [comment]
I opened a separate issue for embedding tests: https://github.com/pythonnet/pythonnet/issues/224

**Comment by matthid** [comment]
Ok I managed to reproduce the problem with the "regular" test suite, but I think it might make more sense to add it to the embedding tests (the code might be a bit simpler there).

I think this PR is now ready to review and merge. What about C#6. Should I still change to C#4?

**Comment by matthid** [comment]
To test that the test cases do indeed reproduce the problems you can use my [add_tests](https://github.com/matthid/pythonnet/tree/add_tests) branch. Then cherry-pick the fixes to make the tests work one by one.

**Comment by denfromufa** [comment]
@matthid travis CI is failing to build your current pull request:

https://travis-ci.org/pythonnet/pythonnet/jobs/135245011

**Comment by matthid** [comment]
Yep investigating, locally its working :(

**Comment by denfromufa** [comment]
> I think this PR is now ready to review and merge. What about C#6. Should I still change to C#4?

No problem, as long as both travis (looks good) and appveyor (have not seen yet) are able to build this.

**Comment by matthid** [comment]
might be a mono bug, an appveyor run might be cool here to clarify

**Comment by matthid** [comment]
Yep its a mono bug:

```csharp
    public class TestConvert : System.Dynamic.DynamicObject
    {
        public override bool TryConvert(System.Dynamic.ConvertBinder binder, out object result)
        {
            result = null;
            return true;
        }
    }

    public class Test : System.Dynamic.DynamicObject
    {
        public override bool TryInvokeMember(System.Dynamic.InvokeMemberBinder binder, object[] args, out object result)
        {
            result = new TestConvert();
            return true;
        }

        public static void TestMethod()
        {
            dynamic t = new Test();
            string result = t.SomeMethod();
        }
    }
```

```csharp
Test.TestMethod(); // Works on .NET, crashes on mono
```

**Comment by matthid** [comment]
We could workaround here at pythonnet by checking for Pythons `None` and returning `null` immediatly in the `TryInvokeMember` call instead of wrapping in a `PyObject` instance. Apparently we cannot convert a `PyObject` to null on mono. Will report this to mono as well but I think we still need a workaround.

**Comment by matthid** [comment]
https://bugzilla.xamarin.com/show_bug.cgi?id=41509

**Comment by matthid** [comment]
Done.

**Comment by vmuriart** [comment]
You can configure a appveyor account to run off your github account. That
what I did for mine to atleast verify my pr

**Comment by matthid** [comment]
@vmuriart Yeah, thanks! But I'm testing on windows already locally. Just mentioned it that it would be nice for others to see (because everything was red above)...

**Comment by tonyroberts** [comment]
I've cherry picked "Alternative Fix for #182" into master. Thanks!

File: src/runtime/pyobject.cs [hunk 3]
```diff
@@ -1133,10 +1133,26 @@ public override bool TryBinaryOperation(BinaryOperationBinder binder, Object arg
...
             return true;
         }
 
+        // Workaround for https://bugzilla.xamarin.com/show_bug.cgi?id=41509
+        // See https://github.com/pythonnet/pythonnet/pull/219
+        private static object CheckNone(PyObject pyObj)
+        {
+            if (pyObj != null)
+            {
+                if (pyObj.obj == Runtime.PyNone)
```
Comment (tonyroberts) [inline src/runtime/pyobject.cs]: Doesn't this alter the behaviour? Previously a PyObject wrapping None would be returned, but now that's replaced with null? I must be missing something...

Comment (matthid) [inline src/runtime/pyobject.cs]: Yeah you are right, it is a workaround that you are not able to return null via dynamic conversion (=later) on mono. please see the linked bug report as well. Feel free to ask if you want a code example (could take some time though)

Comment (tonyroberts) [inline src/runtime/pyobject.cs]: Yeah I get that, but here you *are* returning null where previously an object (PyNone) was returned. Could function just not return pyObj without checking for None? That's the but I don't understand.

Comment (matthid) [inline src/runtime/pyobject.cs]: you could and even should imho, but then you are triggering the mono bug :). The problem is later. when you save such a none object in a dynamic and want to resolve it to a string (for example). This will not work as long as the mono bug exists. In fact one of my tests discovered this glitch

Comment (tonyroberts) [inline src/runtime/pyobject.cs]: Ah ok, that was the bit I was missing - that the bug was triggered later when dealing with the None PyObject instance.

**Comment by matthid** [comment]
> I've cherry picked "Alternative Fix for #182" into master. Thanks!

Ok, but you could have just told me to split the PR or tell me what the problem was with the whole PR (didn't see that there are conflicts until now, because I didn't saw that on the phone).
Fixed the conflicts by rebasing the branch on latest master

**Comment by denfromufa** [comment]
apparently in python 2.7.12 some importing issues have been fixed in C-API:

```
- ``PyImport_Import`` and ``PyImport_ImportModule`` now always do
  absolute imports. In earlier versions they might have used relative
  imports under some conditions.
```

**Comment by matthid** [comment]
@denfromufa I have no idea what you are trying to tell me or what I should do now with this PR :)

**Comment by filmor** [comment]
Okay, decision time: Do we allow C#6? Is there any compelling reason not to, as the tests are working?

In my opinion, this PR can be merged.

**Comment by denfromufa** [comment]
I thought @tonyroberts cherry-picked this pull request?

C# 6.0 is supported by both CI, so I'm +1 on adopting this!

**Comment by matthid** [comment]
@denfromufa Only parts, not sure whats wrong with the rest. I can only improve this when I know whats missing...

**Comment by filmor** [comment]
@denfromufa @vmuriart Merge?

**Comment by denfromufa** [comment]
@filmor can you give me one day to test this locally? I would like to make
sure that it also works with python 2.7.12 due to the bug fix that I linked
above.

On Fri, Jul 22, 2016 at 3:33 AM, Benedikt Reinartz <notifications@github.com

**Comment by denfromufa** [comment]
ok, tested locally and it is working! great finding and fix!

```
C:\Users\denis.akhiyarov>python
Python 2.7.11 |Anaconda 4.0.0 (32-bit)| (default, Mar  4 2016, 15:18:41) [MSC v.
1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import clr
>>> import testimport
>>> from testimport import relative_import

Unhandled Exception: System.ArgumentException: Name must not be empty!
   at Python.Runtime.ModuleObject..ctor(String name)
   at Python.Runtime.ModuleObject.GetAttribute(String name, Boolean guess)
   at Python.Runtime.ImportHook.__import__(IntPtr self, IntPtr args, IntPtr kw)
   at Python.Runtime.Runtime.PyObject_Call(IntPtr pointer, IntPtr args, IntPtr k
w)
   at Python.Runtime.ImportHook.__import__(IntPtr self, IntPtr args, IntPtr kw)

C:\Users\denis.akhiyarov>pip uninstall pythonnet
Uninstalling pythonnet-2.1.0:
  c:\python\python27_32b\lib\site-packages\clr.pyd
  c:\python\python27_32b\lib\site-packages\python.runtime.dll
  c:\python\python27_32b\lib\site-packages\python.runtime.dll.config
  c:\python\python27_32b\lib\site-packages\pythonnet-2.1.0.dist-info\description
.rst
  c:\python\python27_32b\lib\site-packages\pythonnet-2.1.0.dist-info\installer
  c:\python\python27_32b\lib\site-packages\pythonnet-2.1.0.dist-info\metadata
  c:\python\python27_32b\lib\site-packages\pythonnet-2.1.0.dist-info\metadata.js
on
  c:\python\python27_32b\lib\site-packages\pythonnet-2.1.0.dist-info\record
  c:\python\python27_32b\lib\site-packages\pythonnet-2.1.0.dist-info\top_level.t
xt
  c:\python\python27_32b\lib\site-packages\pythonnet-2.1.0.dist-info\wheel
Proceed (y/n)? y
  Successfully uninstalled pythonnet-2.1.0
You are using pip version 8.1.1, however version 8.1.2 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' comm
and.

C:\Users\denis.akhiyarov>pip install git+https://github.com/matthid/pythonnet.gi
t@myfixes
Collecting git+https://github.com/matthid/pythonnet.git@myfixes
  Cloning https://github.com/matthid/pythonnet.git (to myfixes) to c:\users\deni
s~1.akh\appdata\local\temp\pip-inppy5-build
Installing collected packages: pythonnet
  Running setup.py install for pythonnet ... done
Successfully installed pythonnet-2.1.0
You are using pip version 8.1.1, however version 8.1.2 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' comm
and.

C:\Users\denis.akhiyarov>python
Python 2.7.11 |Anaconda 4.0.0 (32-bit)| (default, Mar  4 2016, 15:18:41) [MSC v.
1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import clr
>>> import testimport
>>> from testimport import relative_import
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "testimport\relative_import.py", line 4, in <module>
    from . import _missing_import
  File "testimport\_missing_import.py", line 1, in <module>
    import this_package_should_never_exist_ever
ImportError: No module named this_package_should_never_exist_ever
>>> exit()

C:\Users\denis.akhiyarov>conda update python
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment C:\Python\Python27_32b:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    python-2.7.12              |                0        22.4 MB
    anaconda-custom            |           py27_0          13 KB
    conda-4.1.9                |           py27_0         242 KB
    ------------------------------------------------------------
                                           Total:        22.6 MB

The following packages will be UPDATED:

    anaconda: 4.0.0-np110py27_0 --> custom-py27_0
    conda:    4.1.2-py27_0      --> 4.1.9-py27_0
    python:   2.7.11-4          --> 2.7.12-0

Proceed ([y]/n)? y

Fetching packages ...
python-2.7.12- 100% |###############################| Time: 0:00:14   1.57 MB/s
anaconda-custo 100% |###############################| Time: 0:00:00 293.28 kB/s
conda-4.1.9-py 100% |###############################| Time: 0:00:00 691.15 kB/s
Extracting packages ...
[      COMPLETE      ]|##################################################| 100%
Unlinking packages ...
[      COMPLETE      ]|##################################################| 100%
Linking packages ...
[      COMPLETE      ]|##################################################| 100%

C:\Users\denis.akhiyarov>python
Python 2.7.11 |Anaconda custom (32-bit)| (default, Mar  4 2016, 15:18:41) [MSC v
.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import clr
>>> import testimport
>>> from testimport import relative_import
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "testimport\relative_import.py", line 4, in <module>
    from . import _missing_import
  File "testimport\_missing_import.py", line 1, in <module>
    import this_package_should_never_exist_ever
ImportError: No module named this_package_should_never_exist_ever
>>>
```
